spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.hooks
+
+import org.apache.spark.sql.delta.skipping.clustering.ClusteredTableUtils
+import org.apache.spark.sql.delta._
+import org.apache.spark.sql.delta.actions._
+import org.apache.spark.sql.delta.commands.{DeltaOptimizeContext, OptimizeExecutor}
+import org.apache.spark.sql.delta.commands.optimize._
+import org.apache.spark.sql.delta.metering.DeltaLogging
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+import org.apache.spark.sql.delta.stats.AutoCompactPartitionStats
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.catalog.CatalogTable
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * A trait for post commit hook which compacts files in a Delta table. This hook acts as a cheaper
+ * version of the OPTIMIZE command, by attempting to compact files to a smaller size.
+ *
+ * Auto Compact chooses files to compact greedily by looking at partition directories which
+ * have the largest number of files that are under a certain size threshold and launches a bounded
+ * number of optimize tasks based on the capacity of the cluster.
+ */
+trait AutoCompactBase extends PostCommitHook with DeltaLogging {
+
+  override val name: String = "Auto Compact"
+
+  private[delta] val OP_TYPE = "delta.commit.hooks.autoOptimize"
+
+  /**
+   * This method returns the type of Auto Compaction to use on a delta table or returns None
+   * if Auto Compaction is disabled.
+   * Prioritization:
+   *   1. The highest priority is given to [[DeltaSQLConf.DELTA_AUTO_COMPACT_ENABLED]] config.
+   *   2. Then we check if the deprecated property `DeltaConfigs.AUTO_OPTIMIZE` is set. If yes, then
+   *      we return [[AutoCompactType.Legacy]] type.
+   *   3. Then we check the table property [[DeltaConfigs.AUTO_COMPACT]].
+   *   4. If none of 1/2/3 are set explicitly, then we return None
```

Comment:
It never returns None

Reply:
We actually do. If you look [here](https://github.com/delta-io/delta/pull/2414/files#diff-1506ccd714736b1ce36574203bccc6cc89c66d80b0bdec3b1eb5da6c653e06ccR248) in the `apply` method for `AutoCompactType`, when we pass it `Disabled` it returns `None`